packages/snaps-controllers/src/multichain/MultichainRouter.ts

```diff
@@ -0,0 +1,404 @@
+
+  constructor({ messenger, withSnapKeyring }: MultichainRouterArgs) {
+    this.#messenger = messenger;
+    this.#withSnapKeyring = withSnapKeyring;
+
+    this.#messenger.registerActionHandler(
+      `${name}:handleRequest`,
+      async (...args) => this.handleRequest(...args),
+    );
+
+    this.#messenger.registerActionHandler(
+      `${name}:getSupportedMethods`,
+      (...args) => this.getSupportedMethods(...args),
+    );
+
+    this.#messenger.registerActionHandler(
+      `${name}:getSupportedAccounts`,
+      (...args) => this.getSupportedAccounts(...args),
+    );
+
+    this.#messenger.registerActionHandler(
+      `${name}:isSupportedScope`,
+      (...args) => this.isSupportedScope(...args),
+    );
+  }
+
+  /**
+   * Attempts to resolve the account address to use for a given request by inspecting the request itself.
+   *
+   * The request is sent to to an account Snap via the SnapKeyring that will attempt this resolution.
+   *
+   * @param snapId - The ID of the Snap to send the request to.
+   * @param scope - The CAIP-2 scope for the request.
+   * @param request - The JSON-RPC request.
+   * @returns The resolved address if found, otherwise null.
+   * @throws If the invocation of the SnapKeyring fails.
+   */
+  async #resolveRequestAddress(
+    snapId: SnapId,
+    scope: CaipChainId,
+    request: JsonRpcRequest,
+  ) {
+    try {
+      const result = await this.#withSnapKeyring(async (keyring) =>
+        keyring.resolveAccountAddress(snapId, scope, request),
+      );
+      const address = result?.address;
+      return address ? parseCaipAccountId(address).address : null;
+    } catch {
+      throw rpcErrors.internal();
+    }
+  }
+
+  /**
+   * Get the account ID of the account that should service the RPC request via an account Snap.
+   *
+   * This function checks whether any accounts exist that can service a given request by
+   * using a combination of the resolveAccountAddress functionality and the connected accounts.
+   *
+   * If an account is expected to service this request but none is found, the function will throw.
+   *
+   * @param connectedAddresses - The CAIP-10 addresses connected to the requesting origin.
+   * @param scope - The CAIP-2 scope for the request.
+   * @param request - The JSON-RPC request.
+   * @returns An account ID if found, otherwise null.
+   * @throws If no account is found, but the accounts exist that could service the request.
+   */
+  async #getSnapAccountId(
+    connectedAddresses: CaipAccountId[],
+    scope: CaipChainId,
+    request: JsonRpcRequest,
+  ) {
+    const accounts = this.#messenger
+      .call('AccountsController:listMultichainAccounts', scope)
+      .filter(
+        (
+          account: InternalAccount,
+        ): account is InternalAccount & {
+          metadata: Required<InternalAccount['metadata']>;
+        } =>
+          Boolean(account.metadata.snap?.enabled) &&
+          account.methods.includes(request.method),
```

Comment:
Just a note: this is checked by the `SnapKeyring` before the request is submitted to the Snap, but there is no problem in keeping this check here too.

Reply:
It is important to have it here as we only wanna consider accounts that support the requested method